### PR TITLE
[DataLayout] Refactor parsing of i/f/v/a specifications

### PR DESCRIPTION
**Pull request by s-barannikov**
Split off of #104545 to reduce patch size.

### CODE REVIEW

File: llvm/unittests/IR/DataLayoutTest.cpp [hunk 1]
```diff
@@ -105,6 +72,155 @@ TEST(DataLayout, LayoutStringFormat) {
...
+  std::string format(StringRef Str) const {
+    std::string Res = Str.str();
+    std::replace(Res.begin(), Res.end(), '!', Specifier);
+    return Res;
+  }
+};
+
+INSTANTIATE_TEST_SUITE_P(PrmitiveSpecifiers,
+                         DataLayoutPrimitiveSpecificationTest,
+                         ::testing::Values('i', 'f', 'v'));
```
Comment (s-barannikov) [inline llvm/unittests/IR/DataLayoutTest.cpp:92]: I don't know if this is the right way to do it, but it seems to work as intended.

File: llvm/lib/IR/DataLayout.cpp [hunk 2]
```diff
@@ -376,6 +376,86 @@ static Error getAddrSpace(StringRef R, unsigned &AddrSpace) {
...
+  // ABI alignment. Required, cannot be zero.
+  Align ABIAlign;
+  if (Error Err = parseAlignment(Components[1], ABIAlign, "ABI"))
+    return Err;
+
+  if (Specifier == 'i' && BitWidth == 8 && ABIAlign != 1)
+    return createStringError("i8 must be 8-bit aligned");
+
+  // Preferred alignment. Optional, defaults to the ABI alignment.
+  // Can be zero, meaning use one byte alignment.
```
Comment (nikic) [inline llvm/lib/IR/DataLayout.cpp:403]: This behavior makes little sense to me -- can we remove it?

**Comment by nikic** [comment]
mlir/test/Target/LLVMIR/Import/import-failure.ll is failing.

**Comment by s-barannikov** [comment]
> 	+ ccache --print-stats
>	🚨 Error: The command exited with status 1

Well, I always thought it was a ccache issue. I wonder how many times I could have broken builds by ignoring this error...

**Comment by nikic** [comment]
@s-barannikov It's quite annoying, but what you have to do is download the log and then search for `FAILED:`.

**Comment by s-barannikov** [comment]
> mlir/test/Target/LLVMIR/Import/import-failure.ll is failing.

Apparently, the test relies on successful parsing of invalid specification "i8:8:8:8" by LLVM.
It is then [re-parsed](https://github.com/llvm/llvm-project/blob/90aac06c7f49dd275a49b843b5fd91cb00d549b4/mlir/lib/Target/LLVMIR/ModuleImport.cpp#L617
) by MLIR itself, issuing a different error.

File: mlir/test/Target/LLVMIR/Import/import-failure.ll [hunk 3]
```diff
@@ -360,7 +360,7 @@ declare void @llvm.experimental.noalias.scope.decl(metadata)
 ; // -----
 
 ; CHECK:      import-failure.ll
-; CHECK-SAME: error: cannot translate data layout: i8:8:8:8
+; CHECK-SAME: error: malformed specification
```
Comment (gysit) [inline mlir/test/Target/LLVMIR/Import/import-failure.ll:363]: Feel free to delete this test. It was verifying mlir-translate produces an error if we cannot parse an LLVM data layout string. If the LLVM IR verifier catches this, then there is no reason for this test. Except you know an input that still produces the old error message (e.g. something that is not verified in LLVM but that fails MLIR's import).

Comment (s-barannikov) [reply mlir/test/Target/LLVMIR/Import/import-failure.ll:363]: > If the LLVM IR verifier catches this, then there is no reason for this test.

LLVM should property report any malformed specifications after #104545.

> Except you know an input that still produces the old error message

I tried to find one, but couldn't. I'll delete the test.

Thanks for chiming in quickly!

**Review by nikic — APPROVED** [review]
LGTM

File: mlir/test/Target/LLVMIR/Import/import-failure.ll (hunk 3 above)
Comment (gysit) [reply mlir/test/Target/LLVMIR/Import/import-failure.ll]: Makes sense!

**Pull request merged**